src/guard.rs

```diff
@@ -0,0 +1,160 @@
+    /// Check the size of the given byte slice against a particular type.
+    ///
+    /// # Errors
+    ///
+    /// If the slice's size does not comply with this guard, an error
+    /// which specifies the incompatibility is returned.
+    fn check<T>(v: &[u8]) -> Result<(), GuardError>;
+}
+
+/// Single value guard: The byte slice must have exactly enough bytes to fill a single
+/// instance of a type.
+pub struct SingleValueGuard;
+impl Guard for SingleValueGuard {
+    fn check<T>(bytes: &[u8]) -> Result<(), GuardError> {
+        if bytes.len() != align_of::<T>() {
+            Err(GuardError {
+                required: align_of::<T>(),
+                actual: bytes.len(),
+                reason: ErrorReason::InexactByteCount,
+            })
+        } else {
+            Ok(())
+        }
+    }
+}
+
+/// Pedantic guard: The byte slice must have at least enough bytes to fill a single
+/// instance of a type, and should not have extraneous data.
+pub struct PedanticGuard;
+impl Guard for PedanticGuard {
+    fn check<T>(bytes: &[u8]) -> Result<(), GuardError> {
+        if bytes.len() < align_of::<T>() {
+            Err(GuardError {
+                required: align_of::<T>(),
+                actual: bytes.len(),
+                reason: ErrorReason::NotEnoughBytes,
+            })
+        } else if bytes.len() % align_of::<T>() != 0 {
+            Err(GuardError {
+                required: align_of::<T>(),
+                actual: bytes.len(),
+                reason: ErrorReason::InexactByteCount,
+            })
+        } else {
+            Ok(())
+        }
+    }
+}
+
+/// A strict guard: The byte slice should not have extraneous data, but can be
+/// empty.
+pub struct StrictGuard;
+impl Guard for StrictGuard {
+    fn check<T>(bytes: &[u8]) -> Result<(), GuardError> {
+        if bytes.len() % align_of::<T>() != 0 {
+            Err(GuardError {
+                required: align_of::<T>(),
+                actual: bytes.len(),
+                reason: ErrorReason::InexactByteCount,
+            })
+        } else {
+            Ok(())
+        }
+    }
+}
+
+/// A basic reasonable guard: The byte slice must have at least enough bytes to fill a single
+/// instance of a type, and extraneous data is ignored.
+pub struct BasicGuard;
+impl Guard for BasicGuard {
+    fn check<T>(bytes: &[u8]) -> Result<(), GuardError> {
+        if bytes.len() < align_of::<T>() {
+            Err(GuardError {
+                required: align_of::<T>(),
+                actual: bytes.len(),
+                reason: ErrorReason::NotEnoughBytes,
+            })
+        } else {
+            Ok(())
+        }
+    }
+}
+
+/// Permissive guard: The resulting slice would have as many instances of a type as will
+/// fit, rounded down. Therefore, this guard will never yield an error.
+pub struct PermissiveGuard;
+impl Guard for PermissiveGuard {
+    #[inline]
+    fn check<T>(_bytes: &[u8]) -> Result<(), GuardError> {
+        Ok(())
+    }
+}
```

Comment:
Welcome to pedantic hour with nab.
There's two approaches to this, and I like the former better:

Either slam together struct decls one after another (with a newline in between, we're not heathens), then, slam together impls, separate both blocks by a double newline.

Or keep in this order (interleaved, though I've to say I don't like interleaving decls with impls, spot the C++ programmer), but 1 newline between each decl and its impl and two newlines between each pair.